src/test/java/com/google/sps/DatastoreMetadataStoreTest.java

```diff
@@ -0,0 +1,383 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+
+import static com.google.appengine.api.datastore.FetchOptions.Builder.withLimit;
+
+import java.util.ArrayList;
+import java.util.List;
+import org.junit.Assert;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import com.google.appengine.tools.development.testing.LocalDatastoreServiceTestConfig;
+import com.google.appengine.tools.development.testing.LocalServiceTestHelper;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+
+import com.google.sps.servlets.Metadata;
+import com.google.sps.servlets.DatastoreMetadataStore;
+
+public class DatastoreMetadataStoreTest {
+
+  private final LocalServiceTestHelper helper =
+      new LocalServiceTestHelper(new LocalDatastoreServiceTestConfig());
+
+  // Entities to be put into [mock] Datastore
+  private Entity datasetEntity1;
+  private Entity datasetEntity2;
+  private Entity datasetEntity3;
+  private Entity datasetEntity4;
+  private Entity datasetEntity5;
+  private Entity datasetEntity6;
+  private Entity datasetEntity7;
+  private Entity datasetEntity8;
+  private Entity datasetEntity9;
+  private Entity datasetEntity10;
+
+  // Sample metadata information for an entry in Datastore
+  private static final String USER = "johndallard@google.com";
+  private static final String DATASET = "my_test_dataset";
+  private static final String MODEL = "DELG";
+  private static final String VISUALIZATION = "t-SNE";
+  private static final int NUMBER_OF_IMAGES = 7;
+  private static final long TIMESTAMP = 1596730599767L;
+  
+  @Before
+  public void setUp() {
+    helper.setUp();
+    setUpEntities();
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+  }
+  
+  
+  @Test
+  public void testNumberOfDatastoreEntities() {
+    // Test to show that the correct number of entities end up in datastore
+    
+    DatastoreService ds = DatastoreServiceFactory.getDatastoreService();
+    // Add some entities to Datastore
+    ds.put(datasetEntity1);
```

Comment:
Not sure why you are doing this.

You will never read from the entities created outside your DatastoreMetadataStore library. In fact, you write in this code and read in Python, so maybe you want the opposite test.
Even if you want to test your retrieval with the manually created entity (outside the library), doing it once in basicMetadataRetrieval with a single entity is good enough.
Instead of adding all these redundant lines, you can make your setUp() to do that.
I'd just create several Metadata objects and use datastoreStorage.storeData to write them down and test them.
You probably want to create datastoreStorage in the setUp() and put all those pre-populated data there.

Reply:
Wait it feels like you're asking a few things but I kind of can't make out exactly what you're asking me to change.